tools/Makefile.unix

```diff
@@ -501,6 +501,23 @@ ifeq ($(CONFIG_UBOOT_UIMAGE),y)
 		cp -f uImage /tftpboot/uImage; \
 	fi
 endif
+ifeq ($(CONFIG_ESP32_BINARY),y)
+	@echo "MKIMAGE: ESP32 binary"
+	$(Q) if ! esptool.py version | grep "v2[.]"; then \
+		echo ""; \
+		echo "Please install ESP-IDF tools"; \
+		echo ""; \
+		echo "Check https://docs.espressif.com/projects/esp-idf/en/v4.0/get-started/index.html#installation-step-by-step or run the following command"; \
+		echo ""; \
+		echo "cd tools/esp32 && make && cd ../.."; \
+		echo ""; \
+		echo "run make again to create the nuttx.bin image."; \
+        else \
+		echo "Generating: $(NUTTXNAME).bin (ESP32 compatible)"; \
+		esptool.py --chip esp32 elf2image --flash_mode dio --flash_size 4MB -o nuttx.bin nuttx && \
+		echo "Generated: $(NUTTXNAME).bin (ESP32 compatible)"; \
+	fi
+endif
```

Comment:
CPU specific logic is FORBIDDEN outside of the appropriate arch/ and board/ directories.  This cannot be committed.

Reply:
A more appropriate place to do this would be arch/xtensa/src/Makefile

Reply:
This is required to support a proper modular architecture. If people put compiler-specific, CPU-specific, or any other platform-specific logic outside of those directories, then we damage the architecture. That has never been permitted in the past and and will never be permitted in the future.

There are CPU specific tools under the tools/directory: pic32mx and zds. That is also a precedent for CPU-specific, build-related logic. I am not sure how you integrate it, however.

Reply:
Thanks for the feedback. Yes, I figured out that this code was not good enough to be ready to merge, that is the reason I wanted to work more on it before uploading ... but some people requested some info about how to use QEMU, and it was easier to just create the pull request to share and accelerate the feedback loop. I will try to see if there is a way to move the code to proper location.

Reply:
But now I remember: I did it because it seemed that there was a precedent some lines before in `CONFIG_CXD56_BINARY` (lines 483 to 495). Not that I want to repeat it if that is a "bad practice" or special case/rationale behind it.

Reply:
But It isn't a good place to put vendor specific build step in a common Makefile, it's better to move CONFIG_CXD56_BINARY to arch/arm/src/cxd56 too like Greg's suggestion.

Reply:
> But now I remember: I did it because it seemed that there was a precedent some lines before in `CONFIG_CXD56_BINARY` (lines 483 to 495).

yes, thanks.  I submitted an issue:  #442

Reply:
I wonder if we don't need to hook to run:

    $(TOPDIR)/tools/arch/Makefile at the conclusion of each build.

Where arch is a symbolic link to either a dummy directory or to a directory holding a platform-specific fixup?

Reply:
I do have to fight on an almost daily basis to maintain good modularity in the OS, but I feel like I am losing the battle.  I am being overwhelmed by self-serving, ill-conceived changes that keep the submitter from having do the amount of work that is necessary to maintain the modularity.

Reply:
> I wonder if we don't need to hook to run:
> 
> ```
> $(TOPDIR)/tools/arch/Makefile at the conclusion of each build.
> ```
> 
> Where arch is a symbolic link to either a dummy directory or to a directory holding a platform-specific fixup?

Yes, I was looking at the code to see if something like that was available, but I only saw other architectures were doing directly there.

Reply:
> I do have to fight on an almost daily basis to maintain good modularity in the OS, but I feel like I am losing the battle. I am being overwhelmed by self-serving, ill-conceived changes that keep the submitter from having do the amount of work that is necessary to maintain the modularity.

Sorry to hear that, I don't wanted to cause trouble. But in my defense, if something is not good in the first place, it generates a dilemma on new additions: should we care more about consistency with existing practices (even bad ones) or should we do the correct one despite that will create inconsistencies. The ideal solution should migrate everything, but that puts a burden on the people contributing that was not involved on the original bad practice. So, I will suggest that we should identify that bad practices and at least, if not fix them, mark them as bad _in the code_ (like some FIXME commment). That way the new contributors could be more aware about the preference of other merits over inconsistency in that particular case.

Reply:
> Sorry to hear that, I don't wanted to cause trouble. But in my defense, if something is not good in the first place, it generates a dilemma on new additions: should we care more about consistency with existing practices (even bad ones) or should we do the correct one despite that will create inconsistencies.

Yes, that is exactly the dilemma.  If one bad example is permitted into the architecture, then it provides a precedent and justification for perpetuating the bad example.  The only solution is to be more diligent in review changes in the first place.  But that is not humanly possible, at least not with the people involved now.  Bad things will slip in unnoticed and one they have slipped in, then the problems begin.

I don't have a good solution to that.  But we also can't allow one bad implementation to justify its propagation.

Reply:
Ok, then try to move the other examples of bad practice of this case or at least leave a message in the code if the efforts spans more than a week or so.

I think once there is no "bad example" on place, it would be easier to anyone to point out what is accepted and what is not, so the human review effort would be less, hopefully.

And by the way, from my part, thanks for being so strict :) It pays off on the long term.

I will try to port some of the other cases if I have some time or contribute as needed.

Reply:
PR #457 provides a resolution for the CXD56xx.  If moves the CXD56xx-specific logic out of tools/Makefile.unix&win and into boards/arm/cxd56xx/scripts.  This is the same solution that was used for another architecture so has a solid precedence, works very well, and is very simple.

Reply:
OK.. PR #457 has been merged.  Can you please follow the example there there to implement the POSTBUILD hook for the ESP32.  You only need to look at:

https://github.com/apache/incubator-nuttx/blob/master/boards/arm/cxd56xx/scripts/Config.mk  - New file needed containing the POSTBUILD operations

and

https://github.com/apache/incubator-nuttx/blob/master/boards/arm/cxd56xx/spresense/scripts/Make.defs - Add only like to to include the above file AFTER including tools/Conflg.mk

And that is it!